pkg/sentry/fsimpl/iouringfs/iouringfs.go

```diff
@@ -0,0 +1,318 @@
+)
+
+// filesystemType implements vfs.FilesystemType.
+//
+// +stateify savable
+type iouringFSType struct{}
+
+// GetFilesystem implements vfs.FilesystemType.GetFilesystem.
+func (iouringFSType) GetFilesystem(_ context.Context, vfsObj *vfs.VirtualFilesystem, _ *auth.Credentials, _ string, _ vfs.GetFilesystemOptions) (*vfs.Filesystem, *vfs.Dentry, error) {
+	panic("iouringfs.iouringFSType.GetFilesystem should never be called")
+}
+
+// Name implements vfs.FilesystemType.Name.
+//
+// Note that registering sockfs is unnecessary, except for the fact that it
+// will not show up under /proc/filesystems as a result. This is a very minor
+// discrepancy from Linux.
+func (iouringFSType) Name() string {
+	return "iouring"
+}
+
+// Release implements vfs.FilesystemType.Release.
+func (iouringFSType) Release(ctx context.Context) {}
+
+// +stateify savable
+type iouringFS struct {
+	kernfs.Filesystem
+
+	devMinor uint32
+}
+
+// NewFilesystem sets up and returns a new iouring filesystem.
+//
+// Note that there should only ever be one instance of iouring.Filesystem,
+// backing a global io_uring mount.
+func NewFilesystem(vfsObj *vfs.VirtualFilesystem) (*vfs.Filesystem, error) {
+	devMinor, err := vfsObj.GetAnonBlockDevMinor()
+	if err != nil {
+		return nil, err
+	}
+
+	fs := &iouringFS{
+		devMinor: devMinor,
+	}
+
+	fs.Filesystem.VFSFilesystem().Init(vfsObj, iouringFSType{}, fs)
+
+	return fs.Filesystem.VFSFilesystem(), nil
+}
+
+// Release implements vfs.FilesystemImpl.Release.
+func (fs *iouringFS) Release(ctx context.Context) {
+	fs.Filesystem.VFSFilesystem().VirtualFilesystem().PutAnonBlockDevMinor(fs.devMinor)
+	fs.Filesystem.Release(ctx)
+}
+
+// PrependPath implements vfs.FilesystemImpl.PrependPath.
+func (fs *iouringFS) PrependPath(ctx context.Context, vfsroot, vd vfs.VirtualDentry, b *fspath.Builder) error {
+	inode := vd.Dentry().Impl().(*kernfs.Dentry).Inode().(*inode)
+	b.PrependComponent(fmt.Sprintf("iouring:[%d]", inode.InodeAttrs.Ino()))
+	return vfs.PrependPathSyntheticError{}
+}
+
+// MountOptions implements vfs.FilesystemImpl.MountOptions.
+func (fs *iouringFS) MountOptions() string {
+	return ""
+}
+
+// inode implements kernfs.Inode.
+//
+// +stateify savable
+type inode struct {
+	kernfs.InodeAttrs
+	kernfs.InodeNoopRefCount
+	kernfs.InodeNotDirectory
+	kernfs.InodeNotSymlink
+}
+
+// Open implements kernfs.Inode.Open. iouringfs inode can not be opened
+func (i *inode) Open(ctx context.Context, rp *vfs.ResolvingPath, d *kernfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	return nil, syserror.EPERM
+}
+
+// StatFS implements kernfs.Inode.StatFS.
+func (i *inode) StatFS(ctx context.Context, fs *vfs.Filesystem) (linux.Statfs, error) {
+	return vfs.GenericStatFS(linux.SOCKFS_MAGIC), syserror.EPERM
+}
+
+// NewDentry constructs and returns a iouringfs dentry.
+//
+// Preconditions: mnt.Filesystem() must have been returned by NewFilesystem().
+func NewDentry(ctx context.Context, mnt *vfs.Mount) *vfs.Dentry {
```

Comment:
This function doesn't need to be public.

Reply:
Fixed